File: InfiniteHorizonEx.m
% Infinite horizon stopping problem

%% problem data
k = 20; % grid size
m = 2;
 
X = zeros(k);
X(5,5) = 1; X(17,10) = 1; X(10,15) = 1; % target states

G = zeros(k,k,m); % stage cost
G(:,:,1) = 1; % holding cost
G(5,5,2) = -12; G(17,10,2) = -70; G(10,15,2) = -150; % stopping cost

%% DP algorithm
J = zeros(k,k); % optimal cost functions
pistar = zeros(k,k); % optimal policy

Jplus = rand(k,k); % placeholder for optimal cost update
t = 0; % value iteration counter

% recursion
while norm(Jplus - J) > 1e-8 % value iteration
    t = t+1
    J = Jplus; 
    for i=1:k  % state loop 
        for j=1:k % state loop
            % interior states
            if (1 < i) && (i < k) && (1 < j) && (j < k)
                Ju = zeros(m,1);
                Ju(1) = G(i,j,1) + [1/4 1/4 1/4 1/4]*[J(i+1,j) J(i-1,j) J(i,j+1) J(i,j-1)]';
                Ju(2) = G(i,j,2); 
                [Jplus(i,j), pistar(i,j)] = min(Ju);
            
            % boundary states
            elseif (i == 1) && (1 < j) && (j < k) % top, can't move up
                Ju = zeros(m,1);
                Ju(1) = G(i,j,1) + [1/3 1/3 1/3]*[J(i+1,j) J(i,j+1) J(i,j-1)]';
                Ju(2) = G(i,j,2); 
                [Jplus(i,j), pistar(i,j)] = min(Ju);
                
            elseif (i == k) && (1 < j) && (j < k) % bottom, can't move down
                Ju = zeros(m,1);
                Ju(1) = G(i,j,1) + [1/3 1/3 1/3]*[J(i-1,j) J(i,j+1) J(i,j-1)]';
                Ju(2) = G(i,j,2); 
                [Jplus(i,j), pistar(i,j)] = min(Ju);
                
            elseif (j == 1) && (1 < i) && (i < k) % left, can't move left
                Ju = zeros(m,1);
                Ju(1) = G(i,j,1) + [1/3 1/3 1/3]*[J(i+1,j) J(i-1,j) J(i,j+1)]';
                Ju(2) = G(i,j,2); 
                [Jplus(i,j), pistar(i,j)] = min(Ju);
                
            elseif (j == k) && (1 < i) && (i < k) % right, can't move right
                Ju = zeros(m,1);
                Ju(1) = G(i,j,1) + [1/3 1/3 1/3]*[J(i+1,j) J(i-1,j) J(i,j-1)]';
                Ju(2) = G(i,j,2); 
                [Jplus(i,j), pistar(i,j)] = min(Ju);
                
            elseif (i == 1) && (j == 1) % top left corner
                Ju = zeros(m,1);
                Ju(1) = G(i,j,1) + [1/2 1/2]*[J(i+1,j) J(i,j+1)]';
                Ju(2) = G(i,j,2); 
                [Jplus(i,j), pistar(i,j)] = min(Ju);
                
            elseif (i == 1) && (j == k) % top right corner
                Ju = zeros(m,1);
                Ju(1) = G(i,j,1) + [1/2 1/2]*[J(i+1,j) J(i,j-1)]';
                Ju(2) = G(i,j,2); 
                [Jplus(i,j), pistar(i,j)] = min(Ju);
                
            elseif (i == k) && (j == 1) % bottom left corner
                Ju = zeros(m,1);
                Ju(1) = G(i,j,1) + [1/2 1/2]*[J(i-1,j) J(i,j+1)]';
                Ju(2) = G(i,j,2); 
                [Jplus(i,j), pistar(i,j)] = min(Ju);
                
            elseif (i == k) && (j == k) % bottom right corner
                Ju = zeros(m,1);
                Ju(1) = G(i,j,1) + [1/2 1/2]*[J(i-1,j) J(i,j-1)]';
                Ju(2) = G(i,j,2); 
                [Jplus(i,j), pistar(i,j)] = min(Ju);
            end
        end
    end
end

% plot optimal policy and value function
colormap(flip(hot));
imagesc(pistar);

figure;
% colormap(flip(copper));
surf(J);
% shading interp

File: InfiniteHorizonEx.py
import numpy as np
import matplotlib.pyplot as plt

# Problem data
k = 20  # Grid size
m = 2

X = np.zeros((k, k))
X[4, 4] = 1
X[16, 9] = 1
X[9, 14] = 1  # Target states

G = np.zeros((k, k, m))  # Stage cost
G[:, :, 0] = 1  # Holding cost
G[4, 4, 1] = -120
G[16, 9, 1] = -70
G[9, 14, 1] = -150  # Stopping cost

# DP algorithm
J = np.zeros((k, k))  # Optimal cost functions
pistar = np.zeros((k, k), dtype=int)  # Optimal policy

Jplus = np.random.rand(k, k)  # Placeholder for optimal cost update
t = 0  # Value iteration counter

# Recursion
while np.linalg.norm(Jplus - J) > 1e-8:  # Value iteration
    t += 1
    J = Jplus.copy()
    for i in range(k):
        for j in range(k):
            # Interior states
            if (0 < i < k - 1) and (0 < j < k - 1):
                Ju = np.zeros(m)
                Ju[0] = G[i, j, 0] + (1 / 4) * np.sum([J[i + 1, j], J[i - 1, j], J[i, j + 1], J[i, j - 1]])
                Ju[1] = G[i, j, 1]
                Jplus[i, j] = np.min(Ju)
                pistar[i, j] = np.argmin(Ju)

            # Boundary states
            elif (i == 0) and (0 < j < k - 1):
                Ju = np.zeros(m)
                Ju[0] = G[i, j, 0] + (1 / 3) * np.sum([J[i + 1, j], J[i, j + 1], J[i, j - 1]])
                Ju[1] = G[i, j, 1]
                Jplus[i, j] = np.min(Ju)
                pistar[i, j] = np.argmin(Ju)

            elif (i == k - 1) and (0 < j < k - 1):
                Ju = np.zeros(m)
                Ju[0] = G[i, j, 0] + (1 / 3) * np.sum([J[i - 1, j], J[i, j + 1], J[i, j - 1]])
                Ju[1] = G[i, j, 1]
                Jplus[i, j] = np.min(Ju)
                pistar[i, j] = np.argmin(Ju)

            elif (j == 0) and (0 < i < k - 1):
                Ju = np.zeros(m)
                Ju[0] = G[i, j, 0] + (1 / 3) * np.sum([J[i + 1, j], J[i - 1, j], J[i, j + 1]])
                Ju[1] = G[i, j, 1]
                Jplus[i, j] = np.min(Ju)
                pistar[i, j] = np.argmin(Ju)

            elif (j == k - 1) and (0 < i < k - 1):
                Ju = np.zeros(m)
                Ju[0] = G[i, j, 0] + (1 / 3) * np.sum([J[i + 1, j], J[i - 1, j], J[i, j - 1]])
                Ju[1] = G[i, j, 1]
                Jplus[i, j] = np.min(Ju)
                pistar[i, j] = np.argmin(Ju)

            # Corner states
            elif (i == 0) and (j == 0):
                Ju = np.zeros(m)
                Ju[0] = G[i, j, 0] + (1 / 2) * np.sum([J[i + 1, j], J[i, j + 1]])
                Ju[1] = G[i, j, 1]
                Jplus[i, j] = np.min(Ju)
                pistar[i, j] = np.argmin(Ju)

            elif (i == 0) and (j == k - 1):
                Ju = np.zeros(m)
                Ju[0] = G[i, j, 0] + (1 / 2) * np.sum([J[i + 1, j], J[i, j - 1]])
                Ju[1] = G[i, j, 1]
                Jplus[i, j] = np.min(Ju)
                pistar[i, j] = np.argmin(Ju)

            elif (i == k - 1) and (j == 0):
                Ju = np.zeros(m)
                Ju[0] = G[i, j, 0] + (1 / 2) * np.sum([J[i - 1, j], J[i, j + 1]])
                Ju[1] = G[i, j, 1]
                Jplus[i, j] = np.min(Ju)
                pistar[i, j] = np.argmin(Ju)

            elif (i == k - 1) and (j == k - 1):
                Ju = np.zeros(m)
                Ju[0] = G[i, j, 0] + (1 / 2) * np.sum([J[i - 1, j], J[i, j - 1]])
                Ju[1] = G[i, j, 1]
                Jplus[i, j] = np.min(Ju)
                pistar[i, j] = np.argmin(Ju)

# Plot optimal policy and value function
plt.figure(figsize=(12, 6))
plt.subplot(1, 2, 1)
plt.imshow(pistar, cmap='hot', extent=[0, k - 1, 0, k - 1], origin='lower')
plt.title('Optimal Policy')
plt.colorbar()

plt.subplot(1, 2, 2)
plt.imshow(J, cmap='copper', extent=[0, k - 1, 0, k - 1], origin='lower')
plt.title('Optimal Value Function')
plt.colorbar()
plt.show()
